File: leetcode-04/solutions/python3/0108.py
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        return self.helper(nums, 0, len(nums) - 1)

    def helper(self, nums: List[int], l: int, r: int) -> TreeNode:
        if l > r:
            return None

        mid = (l + r) >> 1
        root = TreeNode(nums[mid])

        root.left = self.helper(nums, l, mid - 1)
        root.right = self.helper(nums, mid + 1, r)

        return root

File: leetcode-04/solutions/cpp/0108.cpp
class Solution {
 public:
  TreeNode* sortedArrayToBST(vector<int>& nums) {
    return helper(nums, 0, nums.size() - 1);
  }

 private:
  TreeNode* helper(vector<int>& nums, int l, int r) {
    if (l > r) return NULL;

    int mid = (l + r) >> 1;
    TreeNode* root = new TreeNode(nums[mid]);

    root->left = helper(nums, l, mid - 1);
    root->right = helper(nums, mid + 1, r);

    return root;
  }
};